File: GLM/Part 4 - Overdispersion/test-overdispersion.py
import pandas as pd
import numpy as np
import statsmodels.api as sm
import matplotlib.pyplot as plt

# Read the data
DF = pd.read_csv("example_data_overdispersed_counts.csv")

# Fit a Poisson GLM
X = sm.add_constant(DF['age'])
y = DF['apples']
GLM = sm.GLM(y, X, family = sm.families.Poisson()).fit()

# A cursory check: Is the deviance ratio close to 1? (No, it is much larger.)
GLM.deviance / GLM.df_resid

# Simulation-based overdispersion test
# Explanation:
#  - Use Monte Carlo simulation of new outcomes, based on the model
#  - Compute the residual deviance for each simulation
#  - The proportion of simulated deviance larger than observed is a p-value

# Observed deviance
obsDev = GLM.deviance

# Sample size, and fitted values
n = DF.shape[0]
yhat = GLM.fittedvalues

# Function to compute simulated deviance (here: Poisson deviance)
def Poisson_deviance(sim_y, yhat):
    return 2 * np.sum(
        np.where(sim_y == 0, 0, sim_y * np.log(sim_y / yhat)) - (sim_y - yhat)
    )

np.random.seed(1234)
MC = 10000
simDev = np.zeros(MC)
for i in range(MC):
    sim_y = np.random.poisson(lam = yhat)
    simDev[i] = Poisson_deviance(sim_y, yhat)

# p-value
np.mean(obsDev < simDev)

plt.hist(simDev, bins = 30, color = 'grey', edgecolor = 'black',
         label = 'Simulated Deviance',
         range = (min(simDev.min(), obsDev), max(simDev.max(), obsDev)))
plt.axvline(obsDev, color = 'red', linewidth = 2, label='Observed Deviance')
plt.xlabel("Deviance")
plt.ylabel("Frequency")
plt.legend()
plt.show()

File: GLM/Part 4 - Overdispersion/test-overdispersion.R
# Read the data
DF <- read.csv("example_data_overdispersed_counts.csv")

# Fit a Poisson GLM
GLM <- glm(apples ~ age, family = "poisson", data = DF)

# A cursory check: Is the deviance ratio close to 1? (No, it is much larger.)
deviance(GLM) / df.residual(GLM)

# Option 1 (recommended): Use the simulation-based approach in DHARMa:
library("DHARMa")
simres <- simulateResiduals(GLM)
testDispersion(simres)

# Option 2: Implement your own (for comparison with Python)

# Explanation:
#  - Use Monte Carlo simulation of new outcomes, based on the model
#  - Compute the residual deviance for each simulation
#  - The proportion of simulated deviance larger than observed is a p-value

# Observed deviance
obsDev <- deviance(GLM)

# Predictor(s), sample size, and fitted values
X <- DF$age
n <- nrow(DF)
yhat <- fitted(GLM)

set.seed(1234)
MC <- 10000
simDev <- numeric(MC)
for(i in 1:MC){
  sim_y     <- rpois(n = n, lambda = yhat)
  simDev[i] <- sum(poisson()$dev.resids(sim_y, yhat, 1))
}

# p-value
mean(obsDev < simDev)

hist(simDev, xlim = range(c(simDev, obsDev)), breaks = 30,
     main = "Simulated (grey) vs observed deviance (red)", xlab = "Deviance")
abline(v = obsDev, col = 2, lwd = 2)
